### PR TITLE
versioning

### LINKED ISSUE #25: versioning
implement the spec from #4

### PR DESCRIPTION
**Pull request by zargot**
fixes #25

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,2 +1,3 @@
 Cerberus==1.3.4
 PyYAML==6.0
+semver==3.0.0.dev3
```
Comment (DougManuel) [inline requirements.txt:3]: Cool. This is a new library for me.

**Review by DougManuel — APPROVED** [review]
I mostly reviewed the unit test code to see if the functionality meets what I expected from the schema and what I know of the users' use cases. 

Looks all good. 
1) If I understand the code, you are validating whether parts.csv has the correct versioning recorded. Great and important. 

Eventually, I envisioned that we would validate the parts list on a pull request to update the parts list. For that, we'd use our validation package, but we'd probably have a separate schema generated to validate the parts list. @zargot and @yulric , were you thinking the same?

2) Thanks for including an update to the .qmd example and also adding a unit test. Exactly what I was hoping for.

I thought that we probably want to include additional .qmd tutorials for different validation features, such as versioning. So, your current `tutorial.qmd` tutorial could be renamed to `getting_started.qmd` (or something like that), and then we'd have `working_with_versioning.qmd` or something like that. There would eventually be more tutorials such as `making_a_custom_shema.qmd`, etc.

What do you think?

Regardless, good work and I am approving this commit. Don't forget to add to documentation when finished.

**Comment by zargot** [comment]
@DougManuel I agree with splitting up the tutorial as soon as we get enough content. I didn't think about validating the parts list, but it sounds like a good idea, unless it's unnecessary? I'll defer to yulric.

**Comment by DougManuel** [comment]
> @DougManuel I agree with splitting up the tutorial as soon as we get enough content. I didn't think about validating the parts list, but it sounds like a good idea, unless it's unnecessary? I'll defer to yulric.

Validating the parts list isn't a priority at this point, but let's keep in mind for the future. You are off to a good start informally validating a few columns within your validation package.

File: src/odm_validation/versions.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
+"""
```
Comment (yulric) [inline src/odm_validation/versions.py:1]: I would move this comment to be close to the _coerce function

**Review by yulric — CHANGES_REQUESTED** [review]
Overall looks good!

* Don't think `datasets` is the right folder to put the parts.csv file, may be confusing as users will expect actual ODM datasets. I think the right spot would be `assets/v2/dictionary/parts.csv`.
* Is there a test for the generate_cerberus_schema function for versioning? Ideally there would be a version 1 test for each validation rule.

File: tools/generate_schemas.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import sys
+from os.path import join, normpath
+from pathlib import Path
+sys.path.append(join(os.path.dirname(os.path.realpath(__file__)), '../src'))
+
+from odm_validation.validation import generate_validation_schema  # noqa:E402
+import odm_validation.utils as utils  # noqa:E402
+
+
+LOG_FILE = 'generate_schemas.log'
```
[the remark this answers is not in the document]
Comment (zargot) [reply tools/generate_schemas.py]: done

File: tools/generate_schemas.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    setup_logging()
+
+    dir = os.path.dirname(os.path.realpath(__file__))
+    asset_dir = join(dir, '../assets')
+    schema_dir = normpath(join(asset_dir, 'validation-schemas'))
+    dataset_dir = normpath(join(asset_dir, 'datasets'))
+
+    print(f'looking for parts in {dataset_dir}')
+    p = Path(dataset_dir)
+    sources = list(p.glob('parts-v*.csv'))
```
[the remark this answers is not in the document]
Comment (zargot) [reply tools/generate_schemas.py]: done

File: specs/module-functions.md [hunk 5]
```diff
@@ -130,30 +130,33 @@ Generates the cerberus schema containing the validation rules to be used with th
...
-                    }
-                }
-            }
-        },
-        "meta": {
-            "partID": "addresses"
-        }
+    "schemaVersion": "1.2.3",
+    "schema": {
+	"addresses": {
```
[the remark this answers is not in the document]
Comment (zargot) [reply specs/module-functions.md]: done

File: specs/module-functions.md [hunk 6]
```diff
@@ -130,30 +130,33 @@ Generates the cerberus schema containing the validation rules to be used with th
...
+			"meta": {
+			    "partID": "addressID",
+			    "addresses": "pK",
+			    "addressesRequired": "mandatory"
+			}
+		    }
+		}
+	    },
+	    "meta": {
+		"partID": "addresses"
```
[the remark this answers is not in the document]
Comment (zargot) [reply specs/module-functions.md]: done

File: specs/module-functions.md [hunk 7]
```diff
@@ -130,30 +130,33 @@ Generates the cerberus schema containing the validation rules to be used with th
...
+    "schemaVersion": "1.2.3",
+    "schema": {
+	"addresses": {
+	    "type": "list",
+	    "schema": {
+		"type": "dict",
+		"schema": {
+		    "addressID": {
+			"required": True,
+			"meta": {
```
[the remark this answers is not in the document]
Comment (zargot) [reply specs/module-functions.md]: done

File: src/odm_validation/validation.py [hunk 8]
```diff
@@ -3,20 +3,33 @@
...
+    package_version: str
+    errors: List[str]
 
+    def valid(self) -> bool:
+        return len(self.errors) == 0
 
-# globals
+
+# public globals
+ODM_LATEST = "2.0.0"
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/validation.py]: done

File: src/odm_validation/versions.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    }
+    ver = Version(**ver)
+    rest = match.string[match.end() :]  # noqa:E203
+    return ver, rest
+
+
+def parse_version(version: str, id='', label='') -> Version:
+    """Returns 1.0.0 when `version` is empty."""
+    if version is None or version == '':
+        warning(f'missing version defaulted to 1.0.0 for "{id}.{label}"')
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/versions.py]: done

File: src/odm_validation/versions.py (hunk 2 above)
Comment (zargot) [reply src/odm_validation/versions.py]: done

**Comment by zargot** [comment]
missing version1* field impl. converting back to draft.

**Review by yulric — CHANGES_REQUESTED** [review]
* Can't find the test for generating the validation schema for missing mandatory column
* Can we also test version 2 validation schema generation? This can be done in another PR.
* From a PR perspective it would be nice if there were clear inputs and ouputs in the tests. For example, in the `invalid_category_impl` function, rather than testing each property individually, doing a deep equal with a previously defined test error report.
* Is there a linting rule to avoid Any? I would like to avoid Any just to make the codebase more clear.
* This is the main comment, although your approach to this feature I think its making the code confusing.  The parts sheet representation in your code is now one of two types, a v1 type which has _original_key and _original_value which is confusing to me and a v2 type which is the normal parts sheet. I think that's confusing because when generating the schema we need to check whether its one type or the other. I think a better approach is generating the schema using the parts sheet as is, and then modifying the schema if the users wants the schema for version 1.

File: src/odm_validation/part_tables.py [hunk 10]
```diff
@@ -143,30 +370,35 @@ def gen_partdata(parts) -> PartData:
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/part_tables.py:297]: i changed a lot, is this still a thing?

File: src/odm_validation/validation.py [hunk 11]
```diff
@@ -3,20 +3,56 @@
...
+from schemas import Schema
+from versions import __version__, parse_version
+
+
+@dataclass(frozen=True)
+class ValidationReport:
+    data_version: str
+    schema_version: str
+    package_version: str
+    errors: List[str]
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/validation.py:28]: is it ok if i create an issue for this instead?

**Review by yulric — CHANGES_REQUESTED** [review]
Ok, great, the code is much more cleaner and easier to follow than before, great job!

1. Majority of comments are on variable naming
2. Can you add descriptions for the functions in the rule_primitives file
3. I think import_dataset should be renamed to import_csv_file
4. The biggest concern with the codebase right now is the typing. There are a lot of implicit Anys that I would like to catch as well as some generic types like list and tuple that don't have their generics filled. I would like to catch these in the linter preferrably so we can fix them, but that's for another PR.

File: src/odm_validation/validation.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/validation.py:28]: done #48

File: tools/generate_schemas.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+def main():
+    dir = os.path.dirname(os.path.realpath(__file__))
+    asset_dir = join(dir, '../assets')
+    schema_dir = normpath(join(asset_dir, 'validation-schemas'))
+    dataset_dir = normpath(join(asset_dir, 'dictionary'))
+
+    info(f'looking for parts in {dataset_dir}')
+    p = Path(dataset_dir)
+    sources = list(p.glob(f'v*/{PARTS_FILENAME}'))
```
[the remark this answers is not in the document]
Comment (zargot) [reply tools/generate_schemas.py]: done, if `parts_files` is ok?

File: tools/generate_schemas.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    if len(sources) == 0:
+        warning('no files found')
+    info(f'writing schemas to {schema_dir}')
+
+    for parts_file in sources:
```
[the remark this answers is not in the document]
Comment (zargot) [reply tools/generate_schemas.py]: isn't it redundant with both file and path?

File: tools/generate_schemas.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        parts_filename = os.path.basename(parts_file)
+        version_regex = f'.+/v(.+)/{PARTS_FILENAME}'
```
[the remark this answers is not in the document]
Comment (zargot) [reply tools/generate_schemas.py]: done, if the new `parts_path_regex` is as good as a comment :)

File: src/odm_validation/part_tables.py [hunk 15]
```diff
@@ -93,111 +196,122 @@ def is_cat(p):
...
+
+def gen_partmap(parts: Dataset) -> PartMap:
+    {get_partID(part): part for part in parts}
+
 
+def partmap(parts) -> PartMap:
+    return {get_partID(part): part for part in parts}
+
+
+def gen_partdata(parts: Dataset, version: Version):
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/part_tables.py]: done

File: src/odm_validation/rule_primitives.py [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+def get_mapped_attr_id(data, attr_id0, ver: Version) -> PartId:
+    if ver.major == 1:
+        return data.mappings[attr_id0][0].id
+    else:
+        return attr_id0
+
+
+def table_items(data, ver):
+    for table_id0, td in data.table_data.items():
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/rule_primitives.py:66]: i would like it to stay version agnostic. my 0 and 1 notation is just an attempt at the math notation `x_1 = f(x_0)`.

File: src/odm_validation/rule_primitives.py [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+def get_table_meta(part, ver: Version) -> Meta:
+    keys = ['partID', 'partType']
+    if ver.major == 1:
+        keys += [pt.V1_LOCATION, pt.V1_TABLE]
+    m: MetaEntry = {k: part[k] for k in keys}
+    return [m]
+
+
+def get_attr_meta(attr: pt.Part, table_id: PartId, ver: Version):
+    keys = ['partID', table_id, pt.table_required_field(table_id)]
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/rule_primitives.py]: done

File: src/odm_validation/rule_primitives.py [hunk 18]
```diff
@@ -0,0 +1,84 @@
+from typing import Dict, List
+
+import part_tables as pt
+from part_tables import Mapping, Meta, MetaEntry, PartId
+from stdext import flatten
+from versions import Version
+
+
+def get_table_meta(part, ver: Version) -> Meta:
+    keys = ['partID', 'partType']
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/rule_primitives.py]: done

File: src/odm_validation/rule_primitives.py [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+    if ver.major == 1:
+        keys += [pt.V1_LOCATION, pt.V1_TABLE]
+    m: MetaEntry = {k: part[k] for k in keys}
+    return [m]
+
+
+def get_attr_meta(attr: pt.Part, table_id: PartId, ver: Version):
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/rule_primitives.py:17]: is it ok if I keep it like this and refactor when a new rule is added that uses a different set of meta fields?

File: tools/generate_schemas.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (yulric) [reply tools/generate_schemas.py]: Well, `parts_files` says to me its a list containing part files whereas parts_file_paths seems explicit

File: tools/generate_schemas.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (yulric) [reply tools/generate_schemas.py]: To me when I see file I think the contents of a file are stored in that variable.

File: src/odm_validation/rule_primitives.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (yulric) [reply src/odm_validation/rule_primitives.py:66]: I would just call it table_id in that case. The 0 in table_id makes it seem like its version 0 of a table_id. Same with attr_id0 variable.

File: tools/generate_schemas.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (zargot) [reply tools/generate_schemas.py]: done

File: tools/generate_schemas.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (zargot) [reply tools/generate_schemas.py]: done

File: src/odm_validation/rule_primitives.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/rule_primitives.py:73]: done

File: src/odm_validation/rule_primitives.py [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+def table_items(data, ver):
+    for table_id0, td in data.table_data.items():
+        table_id1 = get_mapped_table_id(data, table_id0, ver)
+        table = td.part
+        yield (table_id0, table_id1, table)
+
+
+def catset_items(data, ver):
+    """Iterates over all category-set attributes and their associated data."""
+    for attr_id0, cs_data in data.catset_data.items():
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/rule_primitives.py:83]: done (attr_id and doc)

**Pull request merged**